Initial commit of Java Agent for OpenTracing

objectiser opened the pull request:
Initial version of the Java Agent.

The rules for installing opentracing-contrib java integrations will be contained in the agent initially, although have done an experiment with OkHttp3 integration that shows the rule can be included in the integration jar and simply be included in the classpath to have the interceptor automatically installed by the agent. Until the interface used by the rules is stable (may change based on the span management discussion), it is better that the rules are local to the agent.

- [x] Add rule to install okhttp3 interceptor
- [x] Add rule to install servlet filter - Jetty and Tomcat as container specific
- [x] Add rules for instrumenting java.net.HttpUrlConnection (not via integration but directly).
- [x] Tests for adding custom rules in the classpath
- [ ] Add integration test for tomcat servlet rule
- [x] Mechanism to prevent trace data being reported (e.g. when tracer using the tech being instrumented)

NOTE: Added a mechanism to the helper class to keep track of whether a span had been finished. This will be replace by the `Span.isFinished()` method when available.

pavolloffay commented on opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingManager.java:
```diff
@@ -0,0 +1,141 @@
+	private static Logger log = Logger.getLogger(OpenTracingManager.class.getName());
+
+    private static final String RULE_FILE_EXTENSION = ".btm";
+    private static final String ROOT_RULE_FOLDER = "otagent";
```
`byteman` ? as mentioned above.

pavolloffay commented on opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingManager.java:
```diff
@@ -0,0 +1,141 @@
+    private static final String RULE_FILE_EXTENSION = ".btm";
+    private static final String ROOT_RULE_FOLDER = "otagent";
+    private static final String PATH_JAR_SEPARATOR = "!/";
+    private static final String FILE_SCHEME = "file:";
```
this probably won't change ever. in JDK they put it directly to code.

pavolloffay commented on opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingManager.java:
```diff
@@ -0,0 +1,141 @@
+
+                Files.walkFileTree(rules, new SimpleFileVisitor<Path>() {
+
+					@Override
```
There are tabs here

pavolloffay commented on opentracing-agent-tests/pom.xml:
```diff
@@ -0,0 +1,134 @@
+  </dependencies>
+
+  <build>
+    <plugins>
```
deploy plugin to skip this.

pavolloffay commented:
>  Add rules for instrumenting java.net.HttpUrlConnection (not via integration but directly). 

We probably need to be cautious here, what if some framework is internally using directly instrumented classes.

* How can be instrumentation rules enabled/disabled at start time? E.g. would it make sense to have something like `java -javaaget agent.jar -Dinstrument=okhttp,urlconnection,servlet...  -jar` ? And when no specified instrument all..?
* Can be rules added via specifying path as system prop? Can be added agent classpath via system prop too?

objectiser commented:
@pavolloffay Thanks for the comprehensive review.

HttpUrlConnection - have got a solution, which should be available in the next commit - so we can discuss more then. Ideally want to minimize configuration by user, but we can explore different options as problems are identified.

objectiser commented on opentracing-agent-tests/pom.xml:
```diff
@@ -0,0 +1,134 @@
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-failsafe-plugin</artifactId>
+        <configuration>
+          <threadCount>1</threadCount>
```
Don't seem to be necessary currently - so will remove for now.

objectiser commented on opentracing-agent-tests/src/main/scripts/all-assembly.xml:
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
no

objectiser commented on opentracing-agent/src/main/resources/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,5 @@
+Agent-Class: io.opentracing.contrib.agent.OpenTracingAgent
```
will investigate.

objectiser commented on opentracing-agent/src/main/resources/otagent/integrations/okhttp3.btm:
```diff
@@ -0,0 +1,15 @@
+
```
Would prefer to keep the directory name related to the OT agent, in case we need to put other config details in there.

objectiser commented on opentracing-agent-tests/src/test/java/io/opentracing/contrib/agent/tests/OkHttpToServletITest.java:
```diff
@@ -0,0 +1,103 @@
+    public static void initClass() throws Exception {
+        GlobalTracer.register(tracer);
+
+        server = new Server(8180);
```
Will probably change to `okhttp.mockwebserver` - originally was just looking to put together a simple example test to try out the approach - but as seems to be working fine, will separate out into specific test per instrumented framework.

objectiser commented on opentracing-agent-tests/src/test/java/io/opentracing/contrib/agent/tests/OkHttpToServletITest.java:
```diff
@@ -0,0 +1,103 @@
+    public void testRequest() throws IOException {
+        sayHello();
+        assertEquals(2, tracer.finishedSpans().size());
+        assertEquals("GET", tracer.finishedSpans().get(0).operationName());
```
thanks

objectiser commented on opentracing-agent-tests/src/test/java/io/opentracing/contrib/agent/tests/OkHttpToServletITest.java:
```diff
@@ -0,0 +1,103 @@
+        sayHello();
+        assertEquals(2, tracer.finishedSpans().size());
+        assertEquals("GET", tracer.finishedSpans().get(0).operationName());
+        assertEquals("TestSpan", tracer.finishedSpans().get(1).operationName());
```
Agree - this test was initially testing multiple things, including a custom rule - but that can be split out.

objectiser commented on opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingAgent.java:
```diff
@@ -0,0 +1,55 @@
+        } else {
+            agentArgs += ",";
+        }
+        agentArgs += "manager:io.opentracing.contrib.agent.OpenTracingManager";
```
Tried it and it worked! Reason I had it like this was because the javaagent needs to be careful not to pre-emptively cause classes to be loaded before byteman has been initialised - but I think because this OT agent has minimal dependencies it seems to work fine.

objectiser commented on opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingAgent.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    protected static void initProperties() {
+        System.setProperty("org.jboss.byteman.transform.all", "");
```
Needs to be set if byteman is required to change classes in the `java.lang` package. I can remove for now and if we identify such rules, then can be added later.

objectiser commented:
@pavolloffay Haven't addressed the issues you raised yet as wasn't an issue for OkHttp. I think they should be addressed once we have a concrete case where they cause problems - but disabling based on system prop might be an issue as we may need it to be context specific - e.g. app may use httpurlconnection directly in some places, so need tracing, but when used indirectly via some framework then disable. This may be achieved by examining the current span to see if already defined as a `kind=client`, and then not start a new span - but I think it would be better to deal with as a separate PR.

objectiser commented:
@jpkrohling could you review please?

jpkrohling commented on opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingHelper.java:
```diff
@@ -0,0 +1,100 @@
+    private Tracer tracer = GlobalTracer.get();
+    private SpanManager spanManager = DefaultSpanManager.getInstance();
+
+    private static Map<Object,Span> spanCache = Collections.synchronizedMap(new WeakHashMap<Object,Span>());
```
What is the relation between spans from one cache to another? Are they supposed to move from one to another? Can one span be in more than one cache?

objectiser replied:
This 'cache' is establishing an association between an instrumentation target (e.g. a HttpURLConnection) and a span. So its possible multiple application objects will be associated with the same span - it is just to identify the fact that an association has been established and be able to retrieve the associated span.

I think maybe the name should change  - e.g. `spanAssociations`, and method `associateSpan` ?

jpkrohling commented on opentracing-agent/src/main/resources/otagent/direct/java/java-net.btm:
```diff
@@ -0,0 +1,112 @@
+AT EXCEPTION EXIT
+IF span != null && !isFinished($0) && !callerMatches("HttpURLConnection.*",true)
+DO
+  span.setTag(io.opentracing.tag.Tags.HTTP_STATUS.getKey(), Integer.valueOf($0.responseCode))
```
Are you always going to have a `responseCode`? As it's a primitive, it should always be initialize, right?

objectiser replied:
Yes this may be redundant, as recorded in the `AT EXIT` rule - but just realised that it should be recording the `error = true` tag. I'll update.

objectiser replied:
Actually this is still required here as it is triggered on an exceptional exit - i.e. connection refused. But the comment about adding error tag is also true.

jpkrohling commented on pom.xml:
```diff
@@ -0,0 +1,203 @@
+
+  <distributionManagement>
+    <repository>
+      <id>bintray</id>
```
Shouldn't it go to Maven Central?

objectiser replied:
I just copied this from one of the other contrib projects - will double check what the policy is.

pavolloffay replied:
This is from https://github.com/opentracing-contrib/meta/blob/master/guide_java.md. You will need to request several access rights for this(sonatype, bintray). The first release will fail on centra-sync with 403, then you have to go to bintray UI request link to JCenter and when it is approved do the first sync manually.

jpkrohling replied:
What's wrong with deploying to http://oss.sonatype.org/  directly ?

pavolloffay replied:
I don't know, they use the approach described in `ot-java/RELEASE.md`. Zipkin org uses the same approach.. But why I don't know..

objectiser commented on opentracing-agent/src/main/resources/otagent/direct/java/java-net.btm:
```diff
@@ -0,0 +1,112 @@
+#IF span != null && !isFinished($0) && !callerMatches("HttpURLConnection.*",true)
+#		&& $0.responseCode >= 400
+#DO
+#  span.setTag(io.opentracing.tag.Tags.HTTP_STATUS.getKey(), Integer.valueOf($0.responseCode));
```
Same issue here - status code may be redundant, but should record error tag.

objectiser commented:
@jpkrohling Thanks for the comments - have applied some updates if you want to check the latest commit. Realised that I was missing the error tag, so that is now supported, but also that the java.net rules could be simplified to avoid checking the stack.

pavolloffay commented:
Maybe I have missed something but "client" spans should have at least set tag `span.kind=client`

objectiser commented:
@pavolloffay Good point - will add. I assume you mean to the HttpURLConnection rules.

pavolloffay commented:
Yet another thing, what helped me is to use `mockTracer.generatedErrors()` to avoid weird things, but probably in these tests it is not necessary.

objectiser commented:
@jpkrohling When you get a chance, if you could re-review please?

Main changes are that now all of the rules, whether related to directly instrumented techs (e.g. httpurlconnectoin) or integration frameworks, are in separate artifacts - so end user can decide which combination of rules they will use. Eventually the integration framework rules will be moved to the projects themselves.

Although a lot of work still to do - currently it is possible to instrument a simple spring boot app and see the connected output in APM. I just need to add integration test for tomcat, but that will be localised. Then (once reviewed) I think it should be ready to move to opentracing-contrib.

btw - if you could also check the top level README.md, as this will probably be the main documentation for the agent.

jpkrohling commented on .gitignore:
```diff
@@ -0,0 +1,20 @@
+# Eclipse
+.project
+.classpath
+.settings
```
Can you add a `.idea` and `*.iml`, for IntelliJ?

jpkrohling commented on README.md:
```diff
@@ -1,2 +1,205 @@
+
+The Java Agent can be used in two ways:
+
+### Tracer and Framework Integrations on Class Path
```
`s/Class Path/Classpath/` ?

jpkrohling commented on README.md:
```diff
@@ -1,2 +1,205 @@
+
+#### Spring Boot Example
+
+If instrumenting a spring-boot application, add the following to the pom.xml:
```
`s/spring-boot/Spring Boot`

jpkrohling commented on opentracing-agent-itests/src/test/java/io/opentracing/contrib/agent/custom/CustomRuleITest.java:
```diff
@@ -0,0 +1,47 @@
+        assertEquals("OK", spans.get(0).tags().get("status.code"));
+    }
+
+    public void sayHello() {
```
Can you add a comment here, on what is going on behind the scenes? Something short, like:

> When this method is called, the a ByteMan rule should be triggered, creating a "TestSpan".

jpkrohling commented on rules/java-net/src/main/resources/otagent/java-net.btm:
```diff
@@ -0,0 +1,114 @@
+METHOD connect()
+HELPER io.opentracing.contrib.agent.OpenTracingHelper
+AT ENTRY
+IF !$0.connected && retrieveSpan($0) == null && $0.getRequestProperty("opentracing.ignore") == null
```
Can you create a method on the helper, to check for the `opentracing.ignore`? If it's not null, I'd do a boolean check as well: if it's set to `false`, then it should not ignore.

jpkrohling replied:
Another advantage of moving this into the helper is that you could also add a log statement at debug level, like: `log.debug("Ignoring request because the property [opentracing.ignore] is present.")`

objectiser replied:
The problem is what representation should this method take? The HttpURLConnection request properties would be passed in as a `Map<String,List<String>>` -  we couldn't pass in the `HttpURLConection` as this would making the helper mechanism specified/tied to this instrumented type?

jpkrohling replied:
We could, and the method could receive a Object as parameter. Within the method, you could run a couple of "instanceof", and delegate to the appropriate method (if you think it's necessary). I'd say that, as a first implementation, it's quite OK to have it fixed to `HttpUrlConnection`

objectiser replied:
Not sure I like that - but another alternative would be to use an adapter - its the approach OT uses for injecting/extracting state from different transports.

jpkrohling replied:
How about a compromise? A simple solution that solves this problem today + a TODO on the code saying "if this is ever needed for something other than `HttpUrlConnection`, change this code to use adapters".

objectiser replied:
Sounds like a plan :+1:

jpkrohling commented on rules/java-net/src/test/java/io/opentracing/contrib/agent/jnet/JavaHttpURLConnectionITest.java:
```diff
@@ -0,0 +1,223 @@
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+
+public class JavaHttpURLConnectionITest extends OTAgentTestBase {
```
Can you add a javadoc, with a high-level overview on how this test works?

jpkrohling commented on README.md:
```diff
@@ -1,2 +1,205 @@
+
+#### OkHttp
+
+Framework integration repo: [OkHttp](https://github.com/opentracing-contrib/java-okhttp)
```
I'd be curious to know how it would report a HTTP/2 set of requests :)

objectiser replied:
+1 we should add tests if they don't already exist.

objectiser commented:
@jpkrohling Thanks for the review - all updates applied, let me know if there is anything else.

jpkrohling commented on rules/java-web-servlet-filter/src/test/java/io/opentracing/contrib/agent/webservletfilter/TomcatServletITest.java:
```diff
@@ -0,0 +1,105 @@
+public class TomcatServletITest extends OTAgentTestBase {
+
+    // jetty starts on random port
+    private int serverPort = 8080;
```
Will Tomcat always start on this port? Can this be changed, so that it starts at a random high port? This way, if a server already exists at 8080, the test won't fail (like in the Jetty test).

jpkrohling commented:
Apart from the comment on Tomcat's fixed port, this PR looks good to me.

objectiser commented:
@jpkrohling @pavolloffay Thanks for the reviews.
